### PR TITLE
FEAT-#3400: Allow creating a sqlalchemy connection.

### LINKED ISSUE #3400: How to pass the sqlalchemyconnection string for oracle modin read_sql
Comes from https://stackoverflow.com/questions/68999218/how-to-pass-the-sqlalchemyconnection-string-for-oracle-modin-read-sql.

### PR DESCRIPTION
**Pull request by mvashishtha**
Signed-off-by: mvashishtha <mahesh@ponder.io>

## What do these changes do?

With this change, you can have Modin create a sqlalchemy connection
using whatever arguments you want, instead of passing in a
connection string.

I used the new ModinDatabaseConnection to make a psycopg2
connection for a publicly available dataset at
https://rnacentral.org/help/public-database:

```
import modin.pandas as pd
from modin.core.io.sql.connection import ModinDatabaseConnection
con = ModinDatabaseConnection(
    'psycopg2',
    host='hh-pgsql-public.ebi.ac.uk',
    dbname='pfmegrnargs',
    user='reader',
    password='NWDMCE5xdipIjRrp')
df = pd.read_sql("SELECT * FROM rnc_database",
     		 con,
		 index_col=None,
		 coerce_float=True,
		 params=None,
		 parse_dates=None,
		 chunksize=None)
```

Here is how to use the new ModinDatabaseConnection to make a
sqlalchemy connection to the same database:

```
import modin.pandas as pd
from modin.core.io.sql.connection import ModinDatabaseConnection
con = ModinDatabaseConnection(
    'sqlalchemy',
    "postgresql+psycopg2://reader:NWDMCE5xdipIjRrp@hh-pgsql-public.ebi.ac.uk:5432/pfmegrnargs")
df = pd.read_sql("SELECT * FROM rnc_database",
                 con,
                 index_col=None,
                 coerce_float=True,
                 params=None,
                 parse_dates=None,
                 chunksize=None)
```

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3400
- [ ] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/sql/connection.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    lib : str
+        The library for the SQL connection.
+    *args : iterable
+        Positional arguments to pass when creating the connection.
+    **kwargs : dict
+        Keyword arguments to pass when creating the connection.
+    """
+
+    def __init__(self, lib, *args, **kwargs):
+        if lib not in (_PSYCOPG_LIB_NAME, _SQLALCHEMY_LIB_NAME):
```
Comment (devin-petersohn) [inline modin/core/io/sql/connection.py:45]:
```suggestion
        lib = lib.lower()
        if lib not in (_PSYCOPG_LIB_NAME, _SQLALCHEMY_LIB_NAME):
```

File: modin/core/io/sql/connection.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+`ModinDatabaseConnection` saves the arguments that would normally be used to
+make a db connection. It can make and provide the a connection whenever the
+Modin driver or a worker wants one.
+"""
+
+_PSYCOPG_LIB_NAME = "psycopg2"
+_SQLALCHEMY_LIB_NAME = "sqlalchemy"
+
+
+class ModinDatabaseConnection:
```
Comment (devin-petersohn) [inline modin/core/io/sql/connection.py:30]: A top-level API would be a bit better. 

```python
from modin.db_conn import ModinDatabaseConnection
```

Other things will go in that `db_conn` folder like SQL statement construction for specific SQL engines (Like MS SQL) and it feels more user-friendly. What do you think?

File: modin/core/io/sql/sql_dispatcher.py [hunk 3]
```diff
@@ -56,34 +57,24 @@ def _read(cls, sql, con, index_col=None, **kwargs):
...
-                )
-        except ImportError:
-            pass
-        # In the case that we are given a SQLAlchemy Connection or Engine, the objects
-        # are not pickleable. We have to convert it to the URL string and connect from
-        # each of the workers.
-        if not isinstance(con, str):
+        connection_is_modin_db_connection = isinstance(con, ModinDatabaseConnection)
+        if not (connection_is_modin_db_connection or isinstance(con, str)):
             warnings.warn(
```
Comment (devin-petersohn) [inline modin/core/io/sql/sql_dispatcher.py:62]: Should we raise an Exception here?

**Review by devin-petersohn — COMMENTED** [review]
Overall looks good, left a couple of comments.

Would it be possible to add a test for the new functionality?

File: modin/core/io/sql/connection.py (hunk 1 above)
Comment (mvashishtha) [reply modin/core/io/sql/connection.py]: Done.

**Comment by mvashishtha** [comment]
> Would it be possible to add a test for the new functionality?

I added a test checking that we can make a sqlalchemy connection to SQLite.

File: modin/core/io/sql/connection.py (hunk 2 above)
Comment (mvashishtha) [reply modin/core/io/sql/connection.py]: I agree. Done.

File: modin/core/io/sql/sql_dispatcher.py (hunk 3 above)
Comment (mvashishtha) [reply modin/core/io/sql/sql_dispatcher.py:62]: I think it's better to let this case default to Pandas with a warning, as we do now. I prefer to not break existing Pandas workflows with this change.

File: modin/core/io/sql/sql_dispatcher.py [hunk 4]
```diff
@@ -56,34 +57,24 @@ def _read(cls, sql, con, index_col=None, **kwargs):
...
-                "To use parallel implementation of `read_sql`, pass the sqlalchemy"
-                "connection string instead of {}.".format(type(con))
+                "To use parallel implementation of `read_sql`, pass either "
+                "the SQL connection string or a ModinDatabaseConnection "
```
Comment (devin-petersohn) [inline modin/core/io/sql/sql_dispatcher.py:64]: Ideally here we'd point to a webpage where we explain how to use `ModinDatabaseConnection`.

**Review by devin-petersohn — COMMENTED** [review]
It looks like there's a test failure related to the changes here.

Could you also add a page in the readthedocs on this? I think we'll want people to try it right away.

File: modin/core/io/sql/sql_dispatcher.py (hunk 4 above)
Comment (mvashishtha) [reply modin/core/io/sql/sql_dispatcher.py:64]: Done.

**Comment by mvashishtha** [comment]
> It looks like there's a test failure related to the changes here.

Fixed.

> Could you also add a page in the readthedocs on this? I think we'll want people to try it right away.

Done.

File: docs/supported_apis/io_supported.rst [hunk 5]
```diff
@@ -60,3 +60,37 @@ default to pandas.
 .. _`read_sas`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sas.html#pandas.read_sas
 .. _`read_pickle`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_pickle.html#pandas.read_pickle
 .. _`read_sql`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html#pandas.read_sql
+
+Connecting to a database for `read_sql`
+""""""""""""""""""""""""""""""""""""""""""""
```
Comment (devin-petersohn) [inline docs/supported_apis/io_supported.rst:65]: This line should have the same number of characters as the previous line.

File: docs/supported_apis/io_supported.rst [hunk 6]
```diff
@@ -60,3 +60,37 @@ default to pandas.
 .. _`read_sas`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sas.html#pandas.read_sas
 .. _`read_pickle`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_pickle.html#pandas.read_pickle
 .. _`read_sql`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html#pandas.read_sql
+
+Connecting to a database for `read_sql`
```
Comment (devin-petersohn) [inline docs/supported_apis/io_supported.rst:64]: I'd be fine with putting this in the `Using Modin` page instead.

File: docs/supported_apis/io_supported.rst (hunk 5 above)
Comment (mvashishtha) [reply docs/supported_apis/io_supported.rst]: Done.

File: docs/supported_apis/io_supported.rst (hunk 6 above)
Comment (mvashishtha) [reply docs/supported_apis/io_supported.rst]: I moved it there.

File: modin/db_conn.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+in Modin, so each worker must open its own connection.
+`ModinDatabaseConnection` saves the arguments that would normally be used to
+make a db connection. It can make and provide the a connection whenever the
+Modin driver or a worker wants one.
+"""
+
+_PSYCOPG_LIB_NAME = "psycopg2"
+_SQLALCHEMY_LIB_NAME = "sqlalchemy"
+
+
```
Comment (devin-petersohn) [inline modin/db_conn.py:29]:
```suggestion

class UnsupportedDatabaseException(Exception):
    pass


```

**Review by devin-petersohn — COMMENTED** [review]
Looks pretty good, left a couple of nits.

File: modin/db_conn.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    *args : iterable
+        Positional arguments to pass when creating the connection.
+    **kwargs : dict
+        Keyword arguments to pass when creating the connection.
+    """
+
+    def __init__(self, lib, *args, **kwargs):
+        lib = lib.lower()
+        if lib not in (_PSYCOPG_LIB_NAME, _SQLALCHEMY_LIB_NAME):
+            raise Exception("Unsupported database library")
```
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/db_conn.py]: > Can we test this exception?

I also added a test for the new exception.

**Review by devin-petersohn — APPROVED** [review]
LGTM, thanks @mvashishtha !

File: docs/using_modin.rst [hunk 9]
```diff
@@ -63,6 +63,42 @@ If you would like to request a particular method be implemented, feel free to `o
 issue`_. Before you open an issue please make sure that someone else has not already
 requested that functionality.
 
+Connecting to a database for `read_sql`
+---------------------------------------
+
+To make Pandas read from a SQL database, you have two options:
```
Comment (YarShev) [inline docs/using_modin.rst:69]: We are trying to adhere lowercase for `pandas`.

File: modin/core/io/sql/sql_dispatcher.py [hunk 10]
```diff
@@ -56,34 +57,26 @@ def _read(cls, sql, con, index_col=None, **kwargs):
...
-                    else "",  # <port>
-                    con.info.dbname,  # Table in DB
-                )
-        except ImportError:
-            pass
-        # In the case that we are given a SQLAlchemy Connection or Engine, the objects
-        # are not pickleable. We have to convert it to the URL string and connect from
-        # each of the workers.
-        if not isinstance(con, str):
+        connection_is_modin_db_connection = isinstance(con, ModinDatabaseConnection)
```
Comment (YarShev) [inline modin/core/io/sql/sql_dispatcher.py:60]: `connection_is_modin_db_connection` -> `is_modin_db_connection`

File: modin/db_conn.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+"""
+Module houses `ModinDatabaseConnection` class.
+
+`ModinDatabaseConnection` lets a single process make its own connection to a
+database to read from it. Whereas it's possible in pandas to pass an open
+connection directly to `read_sql`, the open connection is not pickleable
+in Modin, so each worker must open its own connection.
+`ModinDatabaseConnection` saves the arguments that would normally be used to
+make a db connection. It can make and provide the a connection whenever the
```
Comment (YarShev) [inline modin/db_conn.py:22]: `the` should be removed

File: docs/using_modin.rst (hunk 9 above)
Comment (mvashishtha) [reply docs/using_modin.rst:69]: Done in #3723.

File: modin/core/io/sql/sql_dispatcher.py (hunk 10 above)
Comment (mvashishtha) [reply modin/core/io/sql/sql_dispatcher.py:60]: Done in #3723.

File: modin/db_conn.py (hunk 11 above)
Comment (mvashishtha) [reply modin/db_conn.py:22]: Done in #3723.

**Pull request merged**